### ISSUE TITLE
decorator fails on NDArray[Any]

### ISSUE THREAD

**Issue by Jasha10** [beartype/beartype#69]
In the [numpy docs](https://numpy.org/doc/stable/reference/typing.html?highlight=ndarray#numpy.typing.NDArray), there is an example showing use of `npt.NDArray[Any]` to type-hint an array with unspecified data type.
Bear type does not seem to support this.

### Reproducing the issue:
```python
# tmp.py
from beartype import beartype
import numpy.typing as npt
from typing import Any


@beartype
def foo(bar: npt.NDArray[Any]) -> Any:
    ...
```
At the console:
```text
$ python tmp.py
Traceback (most recent call last):
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_util/hint/pep/mod/utilmodnumpy.py", line 273, in reduce_hint_numpy_ndarray
    hint_dtype = dtype(hint_dtype_like)
TypeError: Cannot interpret 'typing.Any' as a data type

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jasha10/pysc/tmp.py", line 7, in <module>
    def foo(bar: npt.NDArray[Any]) -> Any:
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_decor/main.py", line 149, in beartype
    func_wrapper_code = generate_code(func_data)
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_decor/_code/codemain.py", line 166, in generate_code
    code_check_params = _code_check_params(data)
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_decor/_code/codemain.py", line 346, in _code_check_params
    hint = sanify_hint_root(
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_util/hint/utilhintconv.py", line 245, in sanify_hint_root
    hint = _reduce_hint(hint, exception_prefix)
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_util/cache/utilcachecall.py", line 351, in _callable_cached
    raise exception
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_util/cache/utilcachecall.py", line 343, in _callable_cached
    return_value = params_flat_to_return_value[params_flat] = func(
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_util/hint/utilhintconv.py", line 697, in _reduce_hint
    hint = reduce_hint_numpy_ndarray(
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/beartype/_util/hint/pep/mod/utilmodnumpy.py", line 281, in reduce_hint_numpy_ndarray
    raise BeartypeDecorHintNonpepNumpyException(
beartype.roar.BeartypeDecorHintNonpepNumpyException: @beartyped foo() parameter "bar" type hint typed NumPy array numpy.ndarray[typing.Any, numpy.dtype[typing.Any]] data type typing.Any invalid (i.e., neither data type nor coercible to data type).
```

**Comment by leycec** [comment]
**Great catch, Jasha.** I shamefully missed that in the NumPy docos – but should have known better. *Everything* is subscriptable by `typing.Any`, right? Right. You know it is.

I'll resolve this straight-away for the global benefit of data scientists everywhere. Thanks again for the heads-up! Double thumbs up to your avatar equipped with stylish coke-bottle glasses, too. :+1: :nerd_face: :+1:

**Comment by leycec** [comment]
**Oh, vile Gods below.** My younger, smarter, more virile self even authored a `FIXME:` comment instructing his older, balder, more RSI-handicapped self to just do this:

``` python
#FIXME: Reduce both the unsubscripted "numpy.typing.NDArray" class *AND* the
#uselessly subscripted "numpy.typing.NDArray[typing.Any]" type hint to the
#"numpy.ndarray" class.
```

It's not a good look. :face_with_head_bandage: 

Relatedly, I'll go ahead and also add support for the unsubscripted `numpy.typing.NDArray` type hint. No one really wants that, but let's avoid any more unseemly issues that make us even balder.

**Comment by leycec** [comment]
**Resolved by 2dbe536.** *It. Is. Done.* Thanks so much for prodding me to get it done too, Jasha. If you're willing to pretend this never happened, I'm willing to pretend I still have hair.

I'd like to kick out a new stable release with this fix in a week or two – but it might be longer. You just never know with @leycec, right? In the meanwhile, all of the following type hints *should* be semantically equivalent to `NDArray[Any]`:

``` python
from beartype import beartype
import numpy as np
import numpy.typing as npt
from typing import Any

# Substitute "Any" -> "np.generic" to make the bear stop growling.
@beartype
def foo(bar: npt.NDArray[np.generic]) -> Any:
    ...

# Substitute "npt.NDArray[Any]" -> "np.ndarray". (This is so lame.)
@beartype
def foo(bar: np.ndarray) -> Any:
    ...
```

I'd probably go with `npt.NDArray[np.generic]`. Gotta preserve those complex type hints, even if they don't actually mean much. I'm sweating profusely for what I have done to your codebase. :sweat_smile:

**Comment by Jasha10** [comment]
Haha, thanks @leycec!
As for the `np.generic` perspiration ... it's ok, that's what search-and-replace is meant for :)
